### DISCUSSION TITLE
please help me understand!

### DISCUSSION THREAD

**Discussion by marioteodoru** [microsoft/WSL#9230] [General]
This is not a issue, it's a question. 

Why is it possible to have  apps like nautilus, or firefox and many of the other WSL2 integrated apps, but i can't  lunch gnome-session?!?!?

**Comment by NotTheDr01ds** [comment]
First, some pointers:

* Keep in mind that I'm not a WSL team member or even Microsoft employee - Just someone hoping to point you in the right direction.
* People will receive your question better if you keep the "rant" out of it.  I'm going to attempt to answer your question, but I'll politely ask you to "tone it down" if you have additional questions or comments.
* It's important to remember that WSL, in its current form at least, is designed primarily to be a tool for *developers*.  It's great that it has been extended to support GUI applications, but running desktop environments is a bit beyond its scope.
* WSL and WSLg are fairly flexible.  If something doesn't work "out of the box" as with `gnome-session`), it's often possible to find a workaround at some level.
* And perhaps most importantly, why assume some nefarious intent?  There are good *technical* reasons that a desktop environment started with `gnome-session` doesn't work "out of the box".

With that in mind, the way that WSLg works (at a very high level) is by providing a Weston implementation (the reference Wayland server) that integrates with the Windows desktop using RDP.  See the [architecture overview](https://devblogs.microsoft.com/commandline/wslg-architecture/) for the deeper explanation.

Weston provides its own window manager, as do all Wayland servers, in my understanding.  When `gnome-session` (or most any desktop environment) tries to start, it also tries to start its window manager (Mutter), and fails because there's already a window-manager running - Weston.  As Steve Pronovost, the WSLg lead, mentions in [this discussion](https://github.com/microsoft/wslg/discussions/81?sort=new) ...

> You should think about WSLg as a desktop environment (really Weston) with individual application integrated back into Windows.

Running a Wayland desktop environment *inside* a Wayland desktop environment just isn't something that is even support under "pure" Linux, that I'm aware of.

Another relevant comment, from Daniel Stone, who is a core Weston/Wayland developer and assisted the WSL team with the WSLg implementation from [this issue](https://bugs.freedesktop.org/show_bug.cgi?id=101741#c6):

> You cannot pick GNOME and a Wayland server separately, they are always a package deal. A desktop environment is a package deal.

That's not to say that this is absolutely *impossible*, just more challenging as a result, and I'm not sure there's a *complete* solution yet.  I *have* been able to bring up Gnome under WSL/WSLg with a bit of a "hack" similar to the one I mentioned for Xfce4 in [this Ask Ubuntu question](https://askubuntu.com/q/1385703/1165986):

```
Xwayland :1 &
WAYLAND_DISPLAY= DISPLAY=:1 dbus-launch gnome-shell --x11
```

This launches an Xwayland backend for Gnome to use, setting it to be display `:1`.  Then it passes the new display number (and an empty `WAYLAND_DISPLAY` to keep Gnome/Mutter from attempting to use it) into the `gnome-shell`.

Without the session manager in place, though, there doesn't appear to be a clean way to "Log off" from Gnome when launched this way.  I think there's a way to configure a particular session to use that particular `gnome-shell` invocation when launched from `gnome-session`, but I'm not sure yet.

This is under Ubuntu-22.04 using WSL 1.0.0 with Systemd enabled.  I do avoid installation of ACPI support, since it seems to cause issues with the Gnome packages:

```
sudo apt-mark hold acpid acpi-support
```

**Comment by I-Want-ToBelieve** [comment]
![image](https://user-images.githubusercontent.com/24669431/205060118-4d6515a5-e3a6-4455-882f-a56f726bfee0.png)
By running x server (VcXsrv) on windows after disabling wslg, I got kde plasma desktop, gnome should work too.
Everything is fine except keyboard hotkey conflicts and graphics card issues

**Reply by NotTheDr01ds** [reply]
@I-Want-ToBelieve  Right - Completely valid approach as well.

The hotkey conflicts, IMHO, are best resolved by running via XRDP so that the keys go directly to the DM, but there's also some [new weirdness](https://askubuntu.com/a/1442818/1165986) that I haven't quite figured out there as well.  It's not quite clear to me why we have to disable WSLg entirely now, but that resolves the problem for Gnome-in-XRDP as well.
